feat: 캘린더 멤버 생성/조회, 캘린더 나가기, 권한 변경

jooyeon0907 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 캘린더 멤버 관련 기능이 구현되지 않음

**TO-BE**
- 캘린더 멤버 생성/조회 기능 구현
- 캘린더 나가기 기능 구현
    - 캘린더를 생성한 사용자가 나가는 경우 (MASTER 권한)
        - 멤버가 본인 제외 1명 이상이라면, 캘린더 멤버 목록 반환
            - 프론트단에서 권한 양도 할 멤버를 선택하기 위함
        - 멤버가 본인뿐이라면, 삭제 요청을 프론트에서 받도록 안내
    - 일반 멤버라면 단순 나가기
- 캘린더 권한 변경 기능 구현

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x] API 테스트 (Postman)
    - 캘린더 멤버 추가 요청
        - JWT 토큰 인증 확인
        - 예외 처리
            - 존재하지 않는 캘린더: "캘린더 정보가 존재하지 않습니다."
            - 이미 초대된 멤버: "해당 캘린더에 이미 초대된 멤버입니다."
        - 캘린더 멤버 추가 성공 여부 확인
    - 캘린더 멤버 조회 요청
        - JWT 토큰 인증 확인
        - 예외 처리
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
        - 캘린더 멤버 조회 성공 여부 확인
    - 캘린더 퇴장 요청
        - JWT 토큰 인증 확인
        - 예외 처리
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 관리자 권한이고, 본인 제외 멤버가 있는 경우: "새로운 캘린더 관리자를 선택해야 합니다."
            - 관리자 권한이고, 멤버가 본인만 남은 경우: "캘린더에 한 명만 남아있는 경우, 해당 캘린더는 삭제됩니다."
        - 일반 멤버인 경우 캘린더 퇴장 성공 여부 확인
    - 캘린더 마스터 권한 변경 요청
        - JWT 토큰 인증 확인
        - 유효성 검사 확인
        - 예외 처리
            - 캘린더 관리자가 아님: "해당 사용자는 이 캘린더의 관리자가 아닙니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 본인에게 권한 양도: "본인에게 권한 양도를 할 수 없습니다."
        - 관리자 권한 양도 및 캘린더 퇴장 성공 여부 확인

jooyeon0907 committed:
feat: 캘린더 멤버 추가/목록조회, 캘린더 나가기, 권한 변경

jooyeon0907 committed:
refactor : 코드 수정

jooyeon0907 committed:
Merge branch 'feature/calendar' of https://github.com/jooyeon0907/timate into feature/calendar_member

 Conflicts:
	src/main/java/com/zerobase/timate/controller/CalendarController.java
	src/main/java/com/zerobase/timate/dto/CalendarDto.java
	src/main/java/com/zerobase/timate/exception/GlobalExceptionHandler.java
	src/main/java/com/zerobase/timate/service/CalendarService.java

jooyeon0907 committed:
Merge branch 'feature/calendar' of https://github.com/jooyeon0907/timate into feature/calendar_member

 Conflicts:
	src/main/java/com/zerobase/timate/service/CommonService.java

jooyeon0907 committed:
refactor: RESTful API 명명 규칙에 맞게 URL 수정

jooyeon0907 committed:
refactor: 코드 리팩토링

jooyeon0907 committed:
refactor: CalendarService 에 캘린더 멤버 체크, 캘린더 MASTER 권한 체크 추가

jooyeon0907 committed:
refactor: 코드 정리 및 수정

jooyeon0907 committed:
fix: 코드 수정

jooyeon0907 committed:
Merge branch 'develop' of https://github.com/jooyeon0907/timate into feature/calendar_member

 Conflicts:
	src/main/java/com/zerobase/timate/controller/UserController.java
	src/main/java/com/zerobase/timate/service/CalendarService.java

jooyeon0907 committed:
refactor: N + 1 문제 발생 가능성으로 Join Fetch 적용

JJae-story commented on src/main/java/com/zerobase/timate/controller/CalendarMemberController.java:
```diff
@@ -0,0 +1,60 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/calendars/{calendarId}/members")
```
안녕하세요!  
@PathVariable의 경로 변수는 항상 메서드에서만 선언해야 하는 줄 알았는데,  
공통된 경로 변수라면 @RequestMapping에서도 지정해서 코드의 중복을 줄일 수 있다는 걸 이번에 알게 됐어요. 감사합니다.

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarMemberService.java:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		UserCalendarId userCalendarId = new UserCalendarId(calendarId, user.getId());
+		UserCalendar userCalendar = new UserCalendar().builder()
+			.id(userCalendarId)
+			.user(user)
+			.calendar(calendar)
+			.role(MemberRole.MEMBER)
+			.build();
```
요런 builder패턴을 이용한 생성자 부분은 UserCalendar 쪽에 메소드로 빼주셔도 좋을 것 같아요. 그러면 UserCalendar에 새로운 필드가 추가되거나 등의 액션이 있을 때 builder 패턴을 사용하는 모든 부분을 찾아서 수정해줄 필요없이 해당 메소드만 수정해주면 편리하거든요. 관심사를 한곳으로 모은다는 측면에서 봐주시면 좋을 것 같아요

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarMemberService.java:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		// 일반 멤버라면 단순 나가기
+		exitCalendar(userId, calendar.getId());
```
여기서 existCalendar를 호출하면 existCalendar에 걸린 @Transactional은 동작하지 않습니다.

https://devgoat.tistory.com/28

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CommonService.java:
```diff
@@ -15,10 +21,73 @@
+	 * @return boolean 사용자가 캘린더의 멤버이면 true, 아니면 false
+	 */
+	public boolean isCalendarMember(Long userId, Long calendarId) {
+		return userCalendarRepository.existsByUserIdAndCalendarId(userId, calendarId);
```
자주 호출된다면 추후 캐싱을 이용해보셔도 좋을 것 같네요

jooyeon0907 replied:
넵 안 그래도 기능을 계속 확장하다보니 자주 호출되는 것을 느끼고 있었는데 캐싱 적용을 해봐야겠어요
이 부분은 새로 브랜치 생성하여 추후 올리도록 하겠습니다!
나머지 피드백은 반영하여 머지 하겠습니다

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CommonService.java:
```diff
@@ -15,10 +21,73 @@
+
+	public void checkCalendarMaster(Long userId, Long calendarId) {
+		UserCalendar userCalendar = getUserCalendar(userId, calendarId);
+		 if (!userCalendar.getRole().equals(MemberRole.MASTER)) {
```
enum의 경우 equals보단 ==로 비교하시면 좋겠습니다

mingming-mentor reviewed (approved):
수고 많으셨어요

jooyeon0907 committed:
fix : 코드 수정
- enum 비교 부분을 equals() 에서 == 로 변경 (Enums 는 자바에서 싱글톤으로 구현되므로 equlas() 비교는 불필요한 객체 비교임)
- builder 패턴을 이용한 생성자 코드를 엔티티 쪽으로 옮김 (추후 유지보수 용이를 위함)
    - Calendar 엔티티로 builder 패턴을 활용한 생성자 코드 이동
    - UserCalendar 엔티티로 builder 패턴을 활용한 생성자 코드 이동
- 불필요한 @Transactional 어노테이션 제거
